logic/transactions/pool.js

```diff
@@ -0,0 +1,922 @@
+	if (limit && limit < r.length) {
+		r.splice(limit);
+	}
+	return r;
+};
+
+/**
+ * Checks sender has enough credit to apply transaction.
+ * @param {transaction} transaction
+ * @param {address} sender
+ * @param {function} cb - Callback function.
+ * @return {setImmediateCallback} err, transactions
+ */
+TransactionPool.prototype.checkBalance  = function (transaction, sender, cb) {
+	var poolBalance = new bignum('0');
+	var paymentTransactions;
+	var receiptTransactions;
+
+	library.logic.account.get({ address: sender.address }, 'balance', function (err, account) {
+		if (err) {
+			return setImmediate(cb, err);
+		}
+		if (account === null) {
+			account = {};
+			account.balance = 0;
+		}
+		// total payments
+		paymentTransactions = self.getAll('sender_id', { id: sender.address });
+		if (paymentTransactions.ready.length > 0) {
+			paymentTransactions.ready.forEach(function (paymentTransaction) {
+				if (paymentTransaction.amount) {
+					poolBalance = poolBalance.minus(paymentTransaction.amount.toString());
+				}
+				poolBalance = poolBalance.minus(paymentTransaction.fee.toString());
+			});
+		}
+		
+		// total receipts
+		receiptTransactions = self.getAll('recipient_id', { id: sender.address });
+		if (receiptTransactions.ready.length > 0) {
+			receiptTransactions.ready.forEach(function (receiptTransaction) {
+				if (receiptTransaction.type === transactionTypes.SEND) {
+					poolBalance = poolBalance.plus(receiptTransaction.amount.toString());
+				}
+			});
+		}
+
+		// total balance
+		var balance = new bignum(account.balance.toString());
+		balance = balance.plus(poolBalance);
+
+		// Check confirmed sender balance
+		var amount = new bignum(transaction.amount.toString()).plus(transaction.fee.toString());
+		var exceeded = balance.lessThan(amount);
+
+		if (exceeded) {
+			return setImmediate(cb, [
+				'Account does not have enough LSK:', sender.address,
+				'balance:', balance.div(Math.pow(10,8))
+			].join(' '));
+		}
+
+		return setImmediate(cb, null, [
+			'balance:', balance.div(Math.pow(10,8))
+		].join(' '));
+	});
+};
+
+/**
+ * Validates a transaction and adds it to the verified pool list.
+ * @implements {__private.addToVerified}
+ * @param {transaction} transaction
+ * @param {boolean} broadcast
+ * @param {function} cb - Callback function.
+ * @return {setImmediateCallback} error | cb
+ */
+TransactionPool.prototype.addFromPublic = function (transaction, broadcast, cb) {
+	__private.addToVerified(transaction, broadcast, cb);
+};
+
+/**
+ * Adds transactions to unverified pool list.
+ * @implements {__private.addToUnverified}
+ * @param {transaction} transactions
+ * @param {boolean} broadcast
+ * @param {function} cb - Callback function.
+ * @return {setImmediateCallback} error | cb
+ */
+TransactionPool.prototype.addFromPeer = function (transactions, broadcast, cb) {
+	if (!Array.isArray(transactions)) {
+		transactions = [transactions];
+	}
+	async.eachSeries(transactions, function (transaction, cb) {
```

Comment:
do we need eachSeries?

Reply:
is in order to get some type of chronological order from transactions that we receive from peers (inserted by receiveAt with milliseconds)